README.md
# rmrebuttal
Use DevStats databases and tools to generate contributors stats

# Data

- Data is generated by single SQL [query](https://github.com/lukaszgryglicki/rmrebuttal/blob/master/query.sql).
- It is processed by `runq` devstats tool and then postprocessed by `rmrebuttal` tool.

# Usage

- You need to have DevStats K8s database and `runq` binary copied to the root directory of this repo.
- Just run `PG_PASS=... ./get_data.sh 10` script to generate Top 10 data for Kubernetes.
- Run `PG_PASS=... PG_DB=prometheus ./get_data.sh 30` script to generate Top 30 data for Prometheus.
- Argument is the N parameter, script analyses Top N developers, contributors, issue creators etc.
- You can also regenerate all data for N=1, 2, 3, 4, 5, 10, 15, 20, 25, 50, 100, 200, 500, 1000 via `PG_PASS=... ./generate_all.sh`.
- You can generate only specific Top N's for a specific project by: `PG_PASS=... NS="1 10 100" PG_DB=prometheus ./generate_all.sh`.
- You can generate data for specific projects (default is all projects) by `PG_PASS=... NS="10 100" PROJS="kubernetes prometheus" ./generate_all.sh`.
- To merge data from multiple Top N analyses, use: `./merge_results.sh`. It will assume default project kubernetes.
- To specify own list of projects use: `PROJS="prometheus allprj" ./merge_results.sh`.
- To select only specific rows use: `RRE="colname;;;regexp" RNAME="name_prefix" ./merge_results.sh`.
- RRE `colname;;;regexp` will be used to select only rows which `colname` matches `regexp`, example:
- `RRE="release;;;(?im)cncf" RNAME="join" PROJS="prometheus allprj" ./merge_results.sh`.
- To have full control on merging use: `PG_DB=projdb ./rmrebuttal nStaticCols 'staticCol1;...;staticColN;dynamicCol1;...;dynamicColN' 'n1;...;N' "colName;;;regexp" output.csv`.
- You can see some examples in `./example_merges.sh`.

# Column format mapping

-  You can specify it by setting COLFMT environment variable as follows:
- `colName1,newName1,type1,fmt1;,,,;colNameN,newNameN,typeN,fmtN`.
- colNameI - required, column name to apply mapping, it can contain '%s' which will be replaced with N if column is dynamic.
- newnameI - new name for column, optional.
- typeI - type of column (to apply format): can be n (numeric), d (datetime), optional.
- fmtI - format of column (if type given), can be Sprintf format for n, or date format for d, optional.
- fmtI for "n" (numeric) column can be for example "%.1f%%".
- fmtI for "d" (datetime) column can be for example: "2006-01-02T15:04:05".
- I = {1,2,...N}.
- Example: `COLFMT="release,Release,,;date_from,Date,d,2006-01-02;top_commits_perc,Percent of top %s committers commits,n,%.1f%%"`.

# Analysis

- Detailed analysis is [here](https://github.com/lukaszgryglicki/rmrebuttal/blob/master/ANALYSIS_RELEASES.md).

# Data sheets and charts

- Top N data and charts are [here](https://docs.google.com/spreadsheets/d/1dK7h8i62G7JEtTrJ2XEYoX0vInEoA7lW0m9ssl5bXag/edit?usp=sharing)
- Contributors data sheets and charts are [here](https://docs.google.com/spreadsheets/d/1EfZEsgfhxUnnOfyHqJO_Uh3aIyDigQkf5bY0AYIrdnE/edit?usp=sharing).
- Committers data sheets and charts are [here](https://docs.google.com/spreadsheets/d/1OY8JHREEZgbvRDFDOplxJJFvEtlcs5TSXbv6JER9_TY/edit?usp=sharing).
- Contributing companies data sheets and charts are [here](https://docs.google.com/spreadsheets/d/1elKts4n5FHzLr9Ob4OUID1MvJ51dmdQDnPP6mgN-O78/edit?usp=sharing).
- Final Kubernetes chart is [here](https://docs.google.com/spreadsheets/d/1tdHqydcDQ7bUWMxHbYADiLmym2UBFcDTUDPCCfRNJuc/edit?usp=sharing).
- Final Prometheus chart is [here](https://docs.google.com/spreadsheets/d/1KtOQRYtflcDVL9FFOJiBxH0adthH5gsIbUPOnFkgIBA/edit?usp=sharing).


releases.txt
2014-01-01 00:00:00 - 2015-07-11 04:02:31:  start - v1.0.0
2015-07-11 04:02:31 - 2015-09-25 23:41:40:  v1.0.0 - v1.1.0
2015-09-25 23:41:40 - 2016-03-16 22:01:03:  v1.1.0 - v1.2.0
2016-03-16 22:01:03 - 2016-07-01 19:19:06:  v1.2.0 - v1.3.0
2016-07-01 19:19:06 - 2016-09-26 18:09:47:  v1.3.0 - v1.4.0
2016-09-26 18:09:47 - 2016-12-12 23:29:43:  v1.4.0 - v1.5.0
2016-12-12 23:29:43 - 2017-03-28 16:23:06:  v1.5.0 - v1.6.0
2017-03-28 16:23:06 - 2017-06-29 22:53:16:  v1.6.0 - v1.7.0
2017-06-29 22:53:16 - 2017-09-28 22:13:57:  v1.7.0 - v1.8.0
2017-09-28 22:13:57 - 2017-12-15 20:53:13:  v1.8.0 - v1.9.0
2017-12-15 20:53:13 - 2018-03-26 16:41:58:  v1.9.0 - v1.10.0
2018-03-26 16:41:58 - 2018-04-11 00:00:00:  v1.10.0 - now


merge_results.sh
#!/bin/bash
if [ -z "$RRE" ]
then
  rname="all"
else
  if [ -z "$RNAME" ]
  then
    echo "$0: if specifying rows by regexp using RRE, you need to specify name prefix too, by using RNAME"
    exit 1
  fi
  rname=$RNAME
fi
if [ -z "$PROJS" ]
then
  PROJS="kubernetes prometheus containerd envoy grpc allprj"
fi
if [ -z "$COLFMT" ]
then
  COLFMT="release,Release,,;date_from,Date from,d,1/2/2006;date_to,Date to,d,1/2/2006;n_top_contributing_coms,Distinct companies for top %s contributors,,;n_top_committing_coms,Distinct companies for top %s committers,,;top_commits_perc,Top %s developers commits as percent of all,n,%.1f%%;top_contributions_perc,Top %s contributors contributions as percent of all,n,%.1f%%;top_comp_contributions_perc,Top %s companies contributions as percent of all,n,%.1f%%;top_comp_commits_perc,Top %s companies commits as percent of all,n,%.1f%%"
fi
export COLFMT
for proj in $PROJS
do
  db=$proj
  if [ "$proj" = "kubernetes" ]
  then
    db="gha"
  elif [ "$proj" = "all" ]
  then
    db="allprj"
  fi
  pref="results/${proj}_${rname}"
  echo "Project: $proj, DB: $db, prefix: $pref"
  PG_DB=$db ./rmrebuttal 3 'release;date_from;date_to;n_top_contributing_coms' '1;2;3;4;5;10;15;20;25;30;50;100;200;500;1000' "$RRE" "${pref}_contributing_companies.csv" || exit 3
  PG_DB=$db ./rmrebuttal 3 'release;date_from;date_to;n_top_committing_coms' '1;2;3;4;5;10;15;20;25;30;50;100;200;500;1000' "$RRE" "${pref}_committing_companies.csv" || exit 4
  PG_DB=$db ./rmrebuttal 3 'release;date_from;date_to;top_contributions_perc' '1;2;3;4;5;10;15;20;25;30;50;100;200;500;1000' "$RRE" "${pref}_percent_contributions.csv" || exit 5
  PG_DB=$db ./rmrebuttal 3 'release;date_from;date_to;top_commits_perc' '1;2;3;4;5;10;15;20;25;30;50;100;200;500;1000' "$RRE" "${pref}_percent_commits.csv" || exit 6
  PG_DB=$db ./rmrebuttal 3 'release;date_from;date_to;top_comp_contributions_perc' '1;2;3;4;5;10;15;20;25;30;50;100;200;500;1000' "$RRE" "${pref}_comp_percent_contributions.csv" || exit 5
  PG_DB=$db ./rmrebuttal 3 'release;date_from;date_to;top_comp_commits_perc' '1;2;3;4;5;10;15;20;25;30;50;100;200;500;1000' "$RRE" "${pref}_comp_percent_commits.csv" || exit 6
done
echo 'OK'


cmd/rmrebuttal/rmrebuttal.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Get cols from results/prefix_top_I.csv files I={n1, n2, ..., nN} and save to ofn.
// Prefix is the project name, like kubernetes, prometheus ... it is taken from PG_DB env variable
func mergeCSVs(stat, cols, ns, rowRegexp, ofn string) error {
	// Static columns, they are not dependent on N
	// Static columns are from the left.
	nStatic, err := strconv.Atoi(stat)
	if err != nil {
		return err
	}

	// Debug/verbose mode
	debug := os.Getenv("DEBUG") != ""

	// Get filename prefix
	prefix := os.Getenv("PG_DB")
	if prefix == "" || prefix == "gha" {
		prefix = "kubernetes"
	}

	// Row regexp handle
	ary := strings.Split(rowRegexp, ";;;")
	lAry := len(ary)
	if lAry > 2 || (lAry == 1 && rowRegexp != "") {
		return fmt.Errorf("'%s' should be einter empty or in 'colname;;;regexp' format", rowRegexp)
	}
	var (
		reColumn  *string
		colRegexp *regexp.Regexp
	)
	if lAry == 2 {
		if ary[0] != "" && ary[1] != "" {
			colRegexp = regexp.MustCompile(ary[1])
			reColumn = &ary[0]
		}
	}

	// column names set
	colsMap := make(map[string]struct{})
	// Column number - to be able to detect if this is a static column
	colNum := make(map[string]int)
	colsAry := strings.Split(cols, ";")
	for i, col := range colsAry {
		if col == "" {
			return fmt.Errorf("empty column definition in '%s'", cols)
		}
		colsMap[col] = struct{}{}
		colNum[col] = i
	}
	lColsAry := len(colsAry)

	// No dynamic columns
	if nStatic >= lColsAry {
		return fmt.Errorf("no dynamic columns, all columns: %d, static columns: %d", lColsAry, nStatic)
	}

	// n values set
	nMap := make(map[int]struct{})
	nAry := strings.Split(ns, ";")
	for _, col := range nAry {
		iCol, err := strconv.Atoi(col)
		if err != nil {
			return err
		}
		nMap[iCol] = struct{}{}
	}

	// Column name mappings, column types and formats
	// "colName1,newName1,type1,fmt1;,,,;colNameN,newNameN,typeN,fmtN"
	// colNameI - required, column name to apply mapping
	// newnameI - new name for column, optional, it can contain '%s' which will be replaced with N if column is dynamic
	// typeI - type of column (to apply format): can be n (numeric), d (datetime), optional
	// fmtI - format of column (if type given), can be Sprintf format for n, or date format for d, optional
	// fmtI for "n" (numeric) column can be for example "%.1f%%".
	// fmtI for "d" (datetime) column can be for example: "2012-11-01T22:08:41+00:00".
	// I = {1,2,...N}
	// Example: COLFMT="release,Release,,;date_from,Date,d,2012-11-01;top_commits_perc,Percent of top %s committers commits,n,%.1f%%"
	colNameMap := make(map[string]string)
	colFmtMap := make(map[string]func(string) string)
	colFmt := os.Getenv("COLFMT")
	if colFmt != "" {
		colFmtAry := strings.Split(colFmt, ";")
		for i, data := range colFmtAry {
			if data == "" {
				return fmt.Errorf("empty column format definition: '%s'", colFmt)
			}
			ary := strings.Split(data, ",")
			lAry := len(ary)
			if lAry != 4 {
				return fmt.Errorf("#%d column format must contain 4 values: '%s', all: '%s'", i, data, colFmt)
			}
			col := ary[0]
			if col == "" {
				return fmt.Errorf("#%d column format must contain column name: '%s', all: '%s'", i, data, colFmt)
			}
			applied := false
			if ary[1] != "" && ary[1] != col {
				colNameMap[col] = ary[1]
				applied = true
			}
			if ary[2] != "" && ary[3] != "" {
				typ := ary[2]
				form := ary[3]
				switch typ {
				case "n":
					colFmtMap[col] = func(in string) string {
						fl, err := strconv.ParseFloat(in, 64)
						if err != nil {
							fmt.Printf("Cannot parse number '%s'\n", in)
							return ""
						}
						if debug {
							fmt.Printf("n_func: form=%s in=%s, fl=%f, out=%s\n", form, in, fl, fmt.Sprintf(form, fl))
						}
						return fmt.Sprintf(form, fl)
					}
					applied = true
				case "d":
					colFmtMap[col] = func(in string) string {
						//tm, e := time.Parse("2006-01-02T15:04:05Z", in)
						tm, err := time.Parse(time.RFC3339, in)
						if err != nil {
							fmt.Printf("Cannot parse datetime '%s'\n", in)
							return ""
						}
						if debug {
							fmt.Printf("d_func: form=%s in=%s, tm=%v, out=%s\n", form, in, tm, tm.Format(form))
						}
						return tm.Format(form)
					}
					applied = true
				default:
					return fmt.Errorf("#%d column contains unknown type specification (allowed: n, d): '%s', all: '%s'", i, data, colFmt)
				}
			}
			if !applied {
				return fmt.Errorf("#%d column contains no usable transformation(s): '%s', all: '%s'", i, data, colFmt)
			}
		}
	}

	// main output: column name --> values
	// each column name ins "columnI J" I-th column and J-th N
	// First nStatic columns doesn not have N added.
	values := make(map[string][]string)
	nN := 0
	for n := range nMap {
		// Read Top N file (current n)
		ifn := fmt.Sprintf("results/%s_top_%d.csv", prefix, n)
		iFile, err := os.Open(ifn)
		if err != nil {
			return err
		}
		defer func() { _ = iFile.Close() }()
		reader := csv.NewReader(iFile)
		rows := 0
		// Will keep column name -> data index map
		colIndex := make(map[string]int)
		for {
			record, err := reader.Read()
			// No more rows
			if err == io.EOF {
				break
			} else if err != nil {
				return err
			}
			rows++
			// Get column -> data index map (from the header row)
			if rows == 1 {
				for i, col := range record {
					_, ok := colsMap[col]
					if ok {
						colIndex[col] = i
					}
				}
				// Check if all columns found
				for col := range colsMap {
					_, ok := colIndex[col]
					if !ok {
						return fmt.Errorf("column '%s' not found in data files", col)
					}
				}
				continue
			}
			// Handle filtering rows by row[reColumn] matching colRegexp
			if reColumn != nil && colRegexp != nil {
				cNum, ok := colNum[*reColumn]
				if !ok {
					return fmt.Errorf("regexp filtering column '%s' not found", *reColumn)
				}
				if cNum >= nStatic {
					return fmt.Errorf(
						"regexp filtering column '%s' is not static (there are %d static cols, this column is #%d)",
						*reColumn,
						nStatic,
						cNum+1,
					)
				}
				index, ok := colIndex[*reColumn]
				if !ok {
					return fmt.Errorf("regexp filtering column '%s' not found by index", *reColumn)
				}
				value := record[index]
				if !colRegexp.MatchString(value) {
					if debug {
						fmt.Printf("Skipping %s=%s, not matching %s\n", *reColumn, value, colRegexp.String())
					}
					continue
				}
			}
			for col, i := range colIndex {
				// Column is "ColName I J" I-th column and J-th N
				// First nStatic columns doesn not have N added.
				colName := col
				cNum := colNum[col]
				if cNum >= nStatic {
					colName = fmt.Sprintf("%s %d", col, n)
				} else {
					// Static column should only be inserted once
					if nN > 0 {
						continue
					}
				}
				_, ok := values[colName]
				if !ok {
					values[colName] = []string{}
				}
				values[colName] = append(values[colName], record[i])
			}
		}
		nN++
	}
	// Write output CSV
	oFile, err := os.Create(ofn)
	if err != nil {
		return err
	}
	defer func() { _ = oFile.Close() }()
	writer := csv.NewWriter(oFile)
	defer writer.Flush()

	// Create header row
	hdr := []string{}
	hdrNoMap := []string{}
	hdrNoN := []string{}
	for _, col := range colsAry {
		name, ok := colNameMap[col]
		// Handle static columns
		cNum := colNum[col]
		if cNum >= nStatic {
			for _, n := range nAry {
				if ok {
					hdr = append(hdr, fmt.Sprintf(name, n))
				} else {
					hdr = append(hdr, fmt.Sprintf("%s %s", col, n))
				}
				hdrNoMap = append(hdrNoMap, fmt.Sprintf("%s %s", col, n))
				hdrNoN = append(hdrNoN, col)
			}
		} else {
			if ok {
				hdr = append(hdr, name)
			} else {
				hdr = append(hdr, col)
			}
			hdrNoMap = append(hdrNoMap, col)
			hdrNoN = append(hdrNoN, col)
		}
	}
	// Write header
	err = writer.Write(hdr)
	if err != nil {
		return err
	}
	// Get length of data and write data
	dataLen := len(values[hdrNoMap[0]])
	for i := 0; i < dataLen; i++ {
		data := []string{}
		for j, col := range hdrNoMap {
			fmtFunc, ok := colFmtMap[hdrNoN[j]]
			if ok {
				data = append(data, fmtFunc(values[col][i]))
			} else {
				data = append(data, values[col][i])
			}
		}
		err = writer.Write(data)
		if err != nil {
			return err
		}
	}
	return nil
}

func main() {
	dtStart := time.Now()
	if len(os.Args) < 6 {
		fmt.Printf("%s: required nStaticCols 'col1;col2;..;colN' 'n1;n2;..;nN' 'colname;;;regexp' output.csv\n", os.Args[0])
		fmt.Printf(
			"Example: %s 3 'date_from;date_to;release;n_top_contributing_coms;top_contributions_perc;"+
				"n_top_committing_coms;top_commits_perc' '10;30;100' 'release;;;(?im)cncf'\n",
			os.Args[0],
		)
		fmt.Printf("%s: use empty colname or empty regexp to skip selecting rows\n", os.Args[0])
		os.Exit(1)
		return
	}
	err := mergeCSVs(os.Args[1], os.Args[2], os.Args[3], os.Args[4], os.Args[5])
	if err != nil {
		fmt.Printf("Error: %s\n", err)
	}
	dtEnd := time.Now()
	fmt.Printf("Time: %v\n", dtEnd.Sub(dtStart))
}
